mmcv/runner/hooks/profiler.py

```diff
@@ -0,0 +1,160 @@
+            operators (matrix multiplication and 2D convolution).
+            Default: False.
+        json_trace_path (str, optional): Exports the collected trace in Chrome
+            JSON format.
+            Default: None.
+
+    Example:
+        >>> runner = ... # instantiate a Runner
+        >>> schedule = dict(wait=1, warmup=1, active=2)
+        >>> # tensorboard trace or log trace
+        >>> trace_config = dict(type='tb_trace', dir_name='work_dir')
+        >>> trace_config = dict(type='log_trace',
+        >>>                     sort_by='self_cuda_time_total',
+        >>>                     row_limit=-1)
+        >>> profiler_config = dict(schedule=schedule,
+        >>>                        on_trace_ready=trace_config)
+        >>> runner.register_profiler_hook(profiler_config)
+        >>> runner.run(data_loaders=[trainloader], workflow=[('train', 1)])
+    """
+
+    def __init__(self,
+                 activities: List[str] = ['cpu', 'cuda'],
+                 schedule: Optional[dict] = None,
+                 on_trace_ready: Optional[Union[Callable, dict]] = None,
+                 record_shapes: bool = False,
+                 profile_memory: bool = False,
+                 with_stack: bool = False,
+                 with_flops: bool = False,
+                 json_trace_path: Optional[str] = None) -> None:
+        try:
+            from torch import profiler  # torch version >= 1.8.0
+        except ImportError:
+            raise ImportError('profiler is the new feature of torch1.8.0, '
+                              f'but your verison is {torch.__version__}')
+
+        if not isinstance(activities, list):
+            raise ValueError(
+                f'activities should be list, but got {type(activities)}')
+        self.activities = []
+        for activity in activities:
+            activity = activity.lower()
+            if activity == 'cpu':
+                self.activities.append(profiler.ProfilerActivity.CPU)
+            elif activity == 'cuda':
+                self.activities.append(profiler.ProfilerActivity.CUDA)
+            else:
+                raise ValueError(
+                    f'activity should be "cpu" or "cuda", but got {activity}')
+
+        if schedule is not None:
+            self.schedule = profiler.schedule(**schedule)
+        else:
+            self.schedule = None
+
+        self.on_trace_ready = on_trace_ready
+        self.record_shapes = record_shapes
+        self.profile_memory = profile_memory
+        self.with_stack = with_stack
+        self.with_flops = with_flops
+        self.json_trace_path = json_trace_path
+
+    @master_only
+    def before_run(self, runner):
+        if callable(self.on_trace_ready):  # handler
+            _on_trace_ready = self.on_trace_ready
+        elif isinstance(self.on_trace_ready, dict):  # config of handler
+            trace_type = self.on_trace_ready.pop('type')  # log_trace handler
```

Comment:
Maybe we should do something like this:
```suggestion
            on_trace_ready = self.on_trace_ready.copy()
            trace_type = on_trace_ready.pop('type')  # log_trace handler
```
Avoiding change the dict in-place.

Reply:
got it